x/incentives/keeper/distribute.go

```diff
@@ -308,29 +320,105 @@ func (k Keeper) AllocateAcrossGauges(ctx sdk.Context) error {
 	return nil
 }
 
-// calcSplitPolicyCoins calculates tokens to split given a policy and groupGauge.
-// TODO: add volume split policy
-// nolint: unused
-func (k Keeper) calcSplitPolicyCoins(policy types.SplittingPolicy, groupGauge *types.Gauge, groupGaugeObj types.GroupGauge) (sdk.Coins, sdk.Coins, error) {
-	if policy == types.Evenly {
-		remainCoins := groupGauge.Coins.Sub(groupGauge.DistributedCoins)
+// syncGroupGaugeWeights updates the individual and total weights of the gauge records based on the splitting policy.
+func (k Keeper) syncGroupGaugeWeights(ctx sdk.Context, groupGauge types.GroupGauge) error {
+	if groupGauge.SplittingPolicy == types.Volume {
+		err := k.syncVolumeSplitGauge(ctx, groupGauge)
+		if err != nil {
+			return err
+		}
+	} else {
+		return types.UnsupportedSplittingPolicyError{GroupGaugeId: groupGauge.GroupGaugeId, SplittingPolicy: groupGauge.SplittingPolicy}
+	}
+
+	return nil
+}
 
-		var coinsDistPerInternalGauge, coinsDistThisEpoch sdk.Coins
-		for _, coin := range remainCoins {
-			epochDiff := groupGauge.NumEpochsPaidOver - groupGauge.FilledEpochs
-			internalGaugeLen := len(groupGaugeObj.InternalIds)
+// syncVolumeSplitGauge syncs a group gauge according to volume splitting policy.
+// It mutates the passed in object and sets the updated value in state.
+// If there is an error, the passed in object is not mutated.
+//
+// It returns an error if:
+// - the splitting policy is not supported
+// - the volume for any linked pool is zero or cannot be found
+// - the cumulative volume for any linked pool has decreased (should never happen)
+func (k Keeper) syncVolumeSplitGauge(ctx sdk.Context, groupGauge types.GroupGauge) error {
+	totalWeight := sdk.ZeroInt()
+
+	// We operate on a deep copy of the given group gauge because we expect to handle specific errors quietly
+	// and want to avoid the scenario where the original group gauge is partially mutated in such cases.
```

Comment:
Since this PR only implements the low level weight syncing functions, errors are bubbled up at this level of abstraction. The quiet handling happens when those functions are run during epoch code (implemented in your downstream PR), while regular handling should happen during gauge creation logic (essentially we should stop people from creating broken groups when possible, but if a group happens to break for a foreseen reason we should not let this lead to a chain halt)